### PR TITLE
Improve whole docker compose stack and add custom stack builder

### PR DESCRIPTION
**Pull request by joshgoebel**
First:

- [x] Fixing local-tooling-webserver https://github.com/exercism/local-tooling-webserver/pull/1 **READY**
- [x] Release new local-tooling-webserver gem **READY**

Checklist for the whole larger Docker niceness project:

- [ ] v3-docker-compose (this PR) **READY**
- [x] Decide on name (startup vs setup)
- [x] update `tooling-invoker` to reference 'exercism/setup') https://github.com/exercism/tooling-invoker/pull/10

Docker Hub:

- [x] website 
- [x] tooling-invoker
- [x] tooling-orchestrator
- [x] setup (v3-docker-compose repo `./setup`) **depends on this PR*
- [x] rebuild `exercism/tooling-invoker` after `exercism/setup` is published
- [ ] javascript_test_runner (dev) **Needs PR below**

Tooling enhancements (shell/tests) for:

- [x] website https://github.com/exercism/v3-website/pull/31 **READY**
- [x] tooling-invoker https://github.com/exercism/tooling-invoker/pull/10 **READY, but will fail to build without this PR)**
- [x] tooling-orchestrator https://github.com/exercism/tooling-orchestrator/pull/10 **READY**
- [ ] test runners? [add sample exercise to test against?](https://github.com/exercism/javascript-test-runner/issues/18)

Runners:

- [ ] Clean up JS test runner to pull both Dockerfiles into JS repo https://github.com/exercism/javascript-test-runner/pull/17 **READY, can have hack removed once local-tooling-webserver is fixed**
- [ ] Clean up CS test runner to pull both Dockerfiles into CS repo 
- [ ] Any nice way to deal with runner vs runner-web-wrapper abstraction?
- [ ] Automate the building of dev.Dockerfile from Dockerfile for runners
- [ ] Any work here to be shared across all runners?

Some of this runner stuff may need to move into a new issue and remain open...

Josh:

- [x] Test the WHOLE stack (submit to marked success)
- [x] Review individual PRs and ping ihid
- [x] Test JS runner
- [x] Test CS runner

### CODE REVIEW

**Comment by joshgoebel** [comment]
Builds on the work in https://github.com/exercism/v3-docker-compose/pull/7

**Comment by joshgoebel** [comment]
```
// before => after
javascript_test_runner_base => javascript_test_runner (just the ./run.sh and tooling)
javascript_test_runner => javascript_test_runner_webhook (with web wrapper)
```

@iHiD Thoughts?

**Comment by iHiD** [comment]
> @iHiD Thoughts? So you have the runner itself, then the runner and webhook.

I think that we should do this slightly differently in general. I think
- We should have a `dev.Dockerfile` within the javascript-test-runner
- When the `Dockerfile` changes, it should create the `dev.Dockerfile` by appending lines to the `Dockerfile`. These could be a new `FROM` block referencing the bit above. This is what gets pushed up to DockerHub.
- The docker compose should just use that dev.Dockerfile.

**Comment by joshgoebel** [comment]
Ok, I completely agree (and then we'd just have a single "service", not a dummy one that starts and fails) but where does this appending/updating happen?  It can't be JUST inside a GHA, the actual file needs to be correct in the repository incase it's needed to be used for a local build as well.

Ok we OK to start by just appending the files manually (to simplify everything else) with a comment marker and then figuring out how to do the auto-update part later?

**Comment by joshgoebel** [comment]
@iHiD Actually that doesn't get us past the naming issue.  There are really TWO runners a person might interact with:

- the "production" one (that's the on that `./run-in-docker` users to allow quick testing) that is just a shell process
- the "development" one that includes the Ruby web wrapper

It's wholly reasonable to hack on the code and then run use `./run-in-docker` to test it... but for the WHOLE stack you'd of course want the web wrapper.

This means someone doing work on test runners likely has GOOD reasons to use BOTH Dockerfile and dev.Dockerfile... I can't think of why we'd want to bloat production with a bunch of Ruby runtimes that aren't needed... so that means we need two names.

I suppose we could do this though:

- exercism/javascript-test-runner (includes web wrapper, for dev like ALL our other packages)
- ???/javascript-test-runner-??? (put the production one in it's own namespace to avoid conflicts with dev)

Thoughts?

**Comment by iHiD** [comment]
So I think the Javascript Test Runner should have a `dev.Dockerfile` which containers the `Dockerfile` with the extra ruby server stuff, and that the image it produces should be suffixed with `-server`, so for example:

```
exercism/javascript-test-runner-server
```

File: docker-compose-full.yml [hunk 1]
```diff
@@ -1,31 +1,53 @@
-version: "3.2"
+version: "2.4"
```
Comment (iHiD) [inline docker-compose-full.yml:1]: Why has this downgraded?

**Comment by iHiD** [comment]
Looking close!!

Two thoughts below:

---

1. Can we add a config setting for daemon, which is set to true by default, but if false just doesn't use the `-d` flag? I would choose to run this not in daemon mode so that the logs stream in the console.

---

2. With reference to the screenshot below, can we check the exit status of the `docker-compose up` to check whether to display the "It may take just a moment for everything to spool up." message?

Screenshot 2020-08-09 at 00 41 56

File: docker-compose-full.yml (hunk 1 above)
Comment (joshgoebel) [reply docker-compose-full.yml:1]: To allow fined control over the processes, and is generally better for docker-compose uses.  If someone wants to do swarm or something more complex than can bump the version manually.

**Comment by joshgoebel** [comment]
> exercism/javascript-test-runner

Don't love it but also not super attached to my ideas.  So yeah it's quite likely that someone use `dev` would not even have runners at all, only runner-servers.

**Comment by joshgoebel** [comment]
> Can we add a config setting for daemon, which is set to true by default, but if false just doesn't use the -d flag? I would choose to run this not in daemon mode so that the logs stream in the console.

Not opposed, but is that a feature request rather than something that should hold up merging? :-)  you can always docker-compose tail the logs.

**Comment by iHiD** [comment]
> is that a feature request rather than something that should hold up merging? :-)

Good question. Probably a feature request, yes.

File: docker-compose-full.yml (hunk 1 above)
Comment (iHiD) [reply docker-compose-full.yml:1]:
```suggestion
# Version 2 allows for more fine control over the processes (memory limits, etc) than version 3.
version: "2.4"
```

File: docker-compose-full.yml [hunk 2]
```diff
@@ -1,31 +1,53 @@
...
 services:
   mysql:
     image: mysql:5.7
+    mem_limit: "1g"
     environment:
       MYSQL_ROOT_PASSWORD: password
-    volumes:
-      - ../v3-website/docker/init.sql:/docker-entrypoint-initdb.d/v3-website.sql
+      MYSQL_ROOT_HOST: '%'
```
Comment (iHiD) [inline docker-compose-full.yml:8]: What does this do?

**Comment by iHiD** [comment]
Running
```
docker build -t exercism/setup --build-arg exercism_config_version=0.38.0 -f setup/dev.Dockerfile ./setup
```

Gives me:
```
...

Successfully installed bundler-2.1.4
1 gem installed
Fetching gem metadata from https://rubygems.org/.........
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using rake 12.3.3
Fetching ast 2.4.1
Installing ast 2.4.1
Fetching aws-eventstream 1.1.0
Installing aws-eventstream 1.1.0
Fetching aws-partitions 1.354.0
Installing aws-partitions 1.354.0
Fetching aws-sigv4 1.2.1
Installing aws-sigv4 1.2.1
Fetching jmespath 1.4.0
Installing jmespath 1.4.0
Fetching aws-sdk-core 3.104.3
Installing aws-sdk-core 3.104.3
Fetching aws-sdk-dynamodb 1.51.0
Installing aws-sdk-dynamodb 1.51.0
Fetching aws-sdk-kms 1.36.0
Installing aws-sdk-kms 1.36.0
Fetching aws-sdk-s3 1.78.0
Installing aws-sdk-s3 1.78.0
Using bundler 2.1.4
Fetching docile 1.3.2
Installing docile 1.3.2
Fetching mandate 0.3.0
Installing mandate 0.3.0
Fetching zeitwerk 2.4.0
Installing zeitwerk 2.4.0
Using exercism-config 0.38.0 from source at `.`
Fetching gem-release 2.1.1
Installing gem-release 2.1.1
Fetching json 2.3.1
Installing json 2.3.1 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /usr/local/bundle/gems/json-2.3.1/ext/json/ext/generator
/usr/local/bin/ruby -I /usr/local/lib/ruby/2.6.0 -r
./siteconf20200811-1-12aj39r.rb extconf.rb
creating Makefile

current directory: /usr/local/bundle/gems/json-2.3.1/ext/json/ext/generator
make "DESTDIR=" clean
sh: make: not found

current directory: /usr/local/bundle/gems/json-2.3.1/ext/json/ext/generator
make "DESTDIR="
sh: make: not found

make failed, exit code 127

Gem files will remain installed in /usr/local/bundle/gems/json-2.3.1 for
inspection.
Results logged to
/usr/local/bundle/extensions/x86_64-linux/2.6.0/json-2.3.1/gem_make.out

An error occurred while installing json (2.3.1), and Bundler cannot continue.
Make sure that `gem install json -v '2.3.1' --source 'https://rubygems.org/'`
succeeds before bundling.

In Gemfile:
  simplecov was resolved to 0.17.1, which depends on
    json
The command '/bin/sh -c gem install bundler:${bundler_version} &&     bundle install' returned a non-zero code: 5
```

**Comment by iHiD** [comment]
Amazing work, @joshgoebel :D

**Pull request merged**